File: src/main/java/maestro/utils/GuiClick.java
package maestro.utils;

import com.mojang.blaze3d.vertex.BufferBuilder;
import com.mojang.blaze3d.vertex.PoseStack;
import java.awt.*;
import java.util.Collections;
import maestro.Agent;
import maestro.api.MaestroAPI;
import maestro.api.pathing.goals.GoalBlock;
import maestro.api.utils.BetterBlockPos;
import maestro.api.utils.Helper;
import maestro.api.utils.MaestroLogger;
import maestro.utils.chat.ChatMessageBuilder;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.client.player.LocalPlayer;
import net.minecraft.core.BlockPos;
import net.minecraft.network.chat.Component;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.level.ClipContext;
import net.minecraft.world.phys.AABB;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.HitResult;
import net.minecraft.world.phys.Vec3;
import org.joml.Matrix4f;
import org.joml.Vector4f;
import org.slf4j.Logger;

public class GuiClick extends Screen implements Helper {
    private static final Logger log = MaestroLogger.get("event");

    private Matrix4f projectionViewMatrix;

    private BlockPos clickStart;
    private BlockPos currentMouseOver;

    public GuiClick() {
        super(Component.literal("CLICK"));
    }

    @Override
    public boolean isPauseScreen() {
        return false;
    }

    @Override
    public void render(GuiGraphics stack, int mouseX, int mouseY, float partialTicks) {
        double mx = mc.mouseHandler.xpos();
        double my = mc.mouseHandler.ypos();

        my = mc.getWindow().getScreenHeight() - my;
        my *= mc.getWindow().getHeight() / (double) mc.getWindow().getScreenHeight();
        mx *= mc.getWindow().getWidth() / (double) mc.getWindow().getScreenWidth();
        Vec3 near = toWorld(mx, my, 0);
        Vec3 far = toWorld(mx, my, 1); // "Use 0.945 that's what stack overflow says" - leijurv

        if (near != null && far != null) {
            Vec3 viewerPos =
                    new Vec3(PathRenderer.posX(), PathRenderer.posY(), PathRenderer.posZ());
            LocalPlayer player =
                    MaestroAPI.getProvider().getPrimaryAgent().getPlayerContext().player();
            BlockHitResult result =
                    player.level()
                            .clip(
                                    new ClipContext(
                                            near.add(viewerPos),
                                            far.add(viewerPos),
                                            ClipContext.Block.OUTLINE,
                                            ClipContext.Fluid.NONE,
                                            player));
            if (result != null && result.getType() == HitResult.Type.BLOCK) {
                currentMouseOver = result.getBlockPos();
            }
        }
    }

    @Override
    public boolean mouseReleased(double mouseX, double mouseY, int mouseButton) {
        if (currentMouseOver
                != null) { // Catch this, or else a click into void will result in a crash
            if (mouseButton == 0) {
                if (clickStart != null && !clickStart.equals(currentMouseOver)) {
                    MaestroAPI.getProvider()
                            .getPrimaryAgent()
                            .getSelectionManager()
                            .removeAllSelections();
                    BetterBlockPos from = BetterBlockPos.from(clickStart);
                    BetterBlockPos to = BetterBlockPos.from(currentMouseOver);
                    MaestroAPI.getProvider()
                            .getPrimaryAgent()
                            .getSelectionManager()
                            .addSelection(from, to);
                    ChatMessageBuilder.info(log, "event")
                            .message("Selection made")
                            .key("from", from)
                            .key("to", to)
                            .withHover("Click to select region")
                            .withClick(
                                    "/maestro sel "
                                            + from.getX()
                                            + " "
                                            + from.getY()
                                            + " "
                                            + from.getZ()
                                            + " "
                                            + to.getX()
                                            + " "
                                            + to.getY()
                                            + " "
                                            + to.getZ())
                            .send();
                    clickStart = null;
                } else {
                    MaestroAPI.getProvider()
                            .getPrimaryAgent()
                            .getCustomGoalProcess()
                            .setGoalAndPath(new GoalBlock(currentMouseOver));
                }
            } else if (mouseButton == 1) {
                MaestroAPI.getProvider()
                        .getPrimaryAgent()
                        .getCustomGoalProcess()
                        .setGoalAndPath(new GoalBlock(currentMouseOver.above()));
            }
        }
        clickStart = null;
        return super.mouseReleased(mouseX, mouseY, mouseButton);
    }

    @Override
    public boolean mouseClicked(double mouseX, double mouseY, int mouseButton) {
        clickStart = currentMouseOver;
        return super.mouseClicked(mouseX, mouseY, mouseButton);
    }

    public void onRender(PoseStack modelViewStack, Matrix4f projectionMatrix) {
        this.projectionViewMatrix = new Matrix4f(projectionMatrix);
        this.projectionViewMatrix.mul(modelViewStack.last().pose());
        this.projectionViewMatrix.invert();

        if (currentMouseOver != null) {
            Entity e = mc.getCameraEntity();
            // drawSingleSelectionBox WHEN?
            PathRenderer.drawManySelectionBoxes(
                    modelViewStack, e, Collections.singletonList(currentMouseOver), Color.CYAN);
            if (clickStart != null && !clickStart.equals(currentMouseOver)) {
                BufferBuilder bufferBuilder =
                        IRenderer.startLines(
                                Color.RED, Agent.settings().pathRenderLineWidthPixels.value, true);
                BetterBlockPos a = new BetterBlockPos(currentMouseOver);
                BetterBlockPos b = new BetterBlockPos(clickStart);
                IRenderer.emitAABB(
                        bufferBuilder,
                        modelViewStack,
                        new AABB(
                                Math.min(a.x, b.x),
                                Math.min(a.y, b.y),
                                Math.min(a.z, b.z),
                                Math.max(a.x, b.x) + 1,
                                Math.max(a.y, b.y) + 1,
                                Math.max(a.z, b.z) + 1));
                IRenderer.endLines(bufferBuilder, true);
            }
        }
    }

    private Vec3 toWorld(double x, double y, double z) {
        if (this.projectionViewMatrix == null) {
            return null;
        }

        x /= mc.getWindow().getWidth();
        y /= mc.getWindow().getHeight();
        x = x * 2 - 1;
        y = y * 2 - 1;

        Vector4f pos = new Vector4f((float) x, (float) y, (float) z, 1.0F);
        projectionViewMatrix.transform(pos);

        if (pos.w() == 0) {
            return null;
        }

        pos.mul(1 / pos.w());
        return new Vec3(pos.x(), pos.y(), pos.z());
    }
}

File: common/src/main/maestro/gui/GuiClick.kt

package maestro.gui

import com.mojang.blaze3d.vertex.BufferBuilder
import com.mojang.blaze3d.vertex.PoseStack
import maestro.Agent
import maestro.api.AgentAPI
import maestro.api.pathing.goals.GoalBlock
import maestro.api.utils.Helper
import maestro.api.utils.Loggers
import maestro.api.utils.PackedBlockPos
import maestro.gui.chat.ChatMessage
import maestro.rendering.IRenderer
import maestro.rendering.PathRenderer
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.screens.Screen
import net.minecraft.core.BlockPos
import net.minecraft.network.chat.Component
import net.minecraft.world.level.ClipContext
import net.minecraft.world.phys.AABB
import net.minecraft.world.phys.HitResult
import net.minecraft.world.phys.Vec3
import org.joml.Matrix4f
import org.joml.Vector4f
import org.slf4j.Logger
import java.awt.Color

class GuiClick :
    Screen(Component.literal("CLICK")),
    Helper {
    private var projectionViewMatrix: Matrix4f? = null
    private var clickStart: BlockPos? = null
    private var currentMouseOver: BlockPos? = null

    override fun isPauseScreen(): Boolean = false

    override fun render(
        stack: GuiGraphics,
        mouseX: Int,
        mouseY: Int,
        partialTicks: Float,
    ) {
        val mc = Minecraft.getInstance()
        val window = mc.window

        var mx = mc.mouseHandler.xpos()
        var my = mc.mouseHandler.ypos()

        my = window.screenHeight - my
        my *= window.height / window.screenHeight.toDouble()
        mx *= window.width / window.screenWidth.toDouble()

        val near = toWorld(mx, my, 0.0)
        val far = toWorld(mx, my, 1.0) // "Use 0.945 that's what stack overflow says" - leijurv

        if (near != null && far != null) {
            val viewerPos = Vec3(PathRenderer.posX(), PathRenderer.posY(), PathRenderer.posZ())
            val player =
                AgentAPI
                    .getProvider()
                    .primaryAgent
                    .playerContext
                    .player()

            val result =
                player.level().clip(
                    ClipContext(
                        near.add(viewerPos),
                        far.add(viewerPos),
                        ClipContext.Block.OUTLINE,
                        ClipContext.Fluid.NONE,
                        player,
                    ),
                )

            if (result != null && result.type == HitResult.Type.BLOCK) {
                currentMouseOver = result.blockPos
            }
        }
    }

    override fun mouseReleased(
        mouseX: Double,
        mouseY: Double,
        mouseButton: Int,
    ): Boolean {
        currentMouseOver?.let { mouseOver ->
            val agent = AgentAPI.getProvider().primaryAgent

            when (mouseButton) {
                0 -> {
                    val start = clickStart
                    if (start != null && start != mouseOver) {
                        agent.selectionManager.removeAllSelections()

                        val from = PackedBlockPos(start)
                        val to = PackedBlockPos(mouseOver)

                        agent.selectionManager.addSelection(from, to)

                        ChatMessage
                            .info(log, "event")
                            .message("Selection made")
                            .key("from", from)
                            .key("to", to)
                            .withHover("Click to select region")
                            .withClick(
                                "/maestro sel ${from.x} ${from.y} ${from.z} ${to.x} ${to.y} ${to.z}",
                            ).send()

                        clickStart = null
                    } else {
                        agent.customGoalTask.setGoalAndPath(GoalBlock(mouseOver))
                    }
                }

                1 -> {
                    agent.customGoalTask.setGoalAndPath(GoalBlock(mouseOver.above()))
                }
            }
        }

        clickStart = null
        return super.mouseReleased(mouseX, mouseY, mouseButton)
    }

    override fun mouseClicked(
        mouseX: Double,
        mouseY: Double,
        mouseButton: Int,
    ): Boolean {
        clickStart = currentMouseOver
        return super.mouseClicked(mouseX, mouseY, mouseButton)
    }

    fun onRender(
        modelViewStack: PoseStack,
        projectionMatrix: Matrix4f,
    ) {
        this.projectionViewMatrix =
            Matrix4f(projectionMatrix).apply {
                mul(modelViewStack.last().pose())
                invert()
            }

        currentMouseOver?.let { mouseOver ->
            val mc = Minecraft.getInstance()
            val e = mc.cameraEntity

            // drawSingleSelectionBox WHEN?
            PathRenderer.drawManySelectionBoxes(
                modelViewStack,
                e,
                listOf(mouseOver),
                Color.CYAN,
            )

            val start = clickStart
            if (start != null && start != mouseOver) {
                val bufferBuilder: BufferBuilder =
                    IRenderer.startLines(
                        Color.RED,
                        Agent.settings().pathRenderLineWidthPixels.value,
                        true,
                    )

                val a = PackedBlockPos(mouseOver)
                val b = PackedBlockPos(start)

                IRenderer.emitAABB(
                    bufferBuilder,
                    modelViewStack,
                    AABB(
                        minOf(a.x, b.x).toDouble(),
                        minOf(a.y, b.y).toDouble(),
                        minOf(a.z, b.z).toDouble(),
                        maxOf(a.x, b.x).toDouble() + 1,
                        maxOf(a.y, b.y).toDouble() + 1,
                        maxOf(a.z, b.z).toDouble() + 1,
                    ),
                )

                IRenderer.endLines(bufferBuilder, true)
            }
        }
    }

    private fun toWorld(
        x: Double,
        y: Double,
        z: Double,
    ): Vec3? {
        val matrix = projectionViewMatrix ?: return null
        val mc = Minecraft.getInstance()

        val normalizedX = x / mc.window.width
        val normalizedY = y / mc.window.height
        val ndcX = normalizedX * 2 - 1
        val ndcY = normalizedY * 2 - 1

        val pos = Vector4f(ndcX.toFloat(), ndcY.toFloat(), z.toFloat(), 1.0f)
        matrix.transform(pos)

        if (pos.w() == 0f) {
            return null
        }

        pos.mul(1 / pos.w())
        return Vec3(pos.x().toDouble(), pos.y().toDouble(), pos.z().toDouble())
    }

    companion object {
        private val log: Logger = Loggers.get("event")
    }
}
